Add PLBart

gchhablani opened the pull request:
# What does this PR do?
This PR adds PLBART.
- Paper: https://arxiv.org/abs/2103.06333
- Code and Checkpoints: https://github.com/wasiahmad/PLBART
- Authors: @wasiahmad @kaiweichang

Motivation: This encoder-decoder BART-like model allows for downstream fine-tuning on code summarization, generation, classification, etc. and is useful to the community. The fine-tuned checkpoints are also available and can be used directly.

EDIT 1:
------
I am trying to make the embeddings same for the original and my implementation. Have created an issue [here](#13481) as I'm stuck with `padding_idx` outputs not matching.

### Pending Tasks
- [ ] Fix FastTokenizer
    - [ ] Tokenizer
    - ~MultiTokenizer~ 
- [x] Fix tests
    - [x] Modeling
    - [x] Tokenizer
    - ~MultiTokenizer~
- [ ] Verify behavior for all checkpoints
- [x] Update Docs
- [ ] Update Model Cards
- [x] Add remaining checkpoints
    - [x] `plbart-large`
    - [x] `plbart-base-csnet` checkpoints

gchhablani committed:
Init PLBART

gchhablani committed:
Add missing configuration file

gchhablani committed:
Add conversion script and configurationf ile

gchhablani committed:
Fix style

gchhablani committed:
Update modeling and conversion scripts

gchhablani committed:
Fix scale embedding in config

gchhablani committed:
Add comment

gchhablani commented:
Unstale

gchhablani committed:
Fix conversion script

gchhablani committed:
Add classification option to conversion script

gchhablani committed:
Fix vocab size in config doc

gchhablani committed:
Add tokenizer files from MBart50

gchhablani committed:
Allow no lang code in regular tokenizer

gchhablani committed:
Add PLBart Tokenizer Converters

gchhablani committed:
Remove mask from multi tokenizer

gchhablani committed:
Remove mask from multi tokenizer

gchhablani committed:
Change from MBart-50 to MBart tokenizer

gchhablani committed:
Fix names and modify src/tgt behavior

gchhablani committed:
Fix imports for tokenizer

gchhablani committed:
Remove <mask> from multi tokenizer

gchhablani committed:
Fix style

gchhablani committed:
Change tokenizer_class to processor_class

gchhablani committed:
Add attribute map to config class

gchhablani committed:
Update modeling file to modified MBart code

gchhablani committed:
Update configuration file to MBart style configuration

gchhablani committed:
Fix tokenizer

gchhablani committed:
Separate tokenizers

gchhablani committed:
Fix error in tokenization auto

gchhablani committed:
Copy MBart tests

gchhablani committed:
Replace with MBart tokenization tests

gchhablani committed:
Fix style

gchhablani committed:
Fix language code in multi tokenizer

gchhablani committed:
Fix configuration docs

gchhablani committed:
Add entry for plbart_multi in transformers init

gchhablani committed:
Add dummy objects and fix imports

gchhablani committed:
Fix modeling tests

gchhablani committed:
Add TODO in config

gchhablani committed:
Fix copyright year

gchhablani committed:
Fix modeling docs and test

gchhablani committed:
Fix some tokenization tests and style

patrickvonplaten commented on docs/source/model_doc/plbart.rst:
```diff
@@ -0,0 +1,89 @@
+.. 
+    Copyright 2020 The HuggingFace Team. All rights reserved.
```
```suggestion
    Copyright 2021 The HuggingFace Team. All rights reserved.
```

patrickvonplaten commented on src/transformers/__init__.py:
```diff
@@ -376,6 +378,8 @@
     _import_structure["models.mbart50"].append("MBart50Tokenizer")
     _import_structure["models.mt5"].append("MT5Tokenizer")
     _import_structure["models.pegasus"].append("PegasusTokenizer")
+    _import_structure["models.plbart"].append("PLBartTokenizer")
```
Do we really need two tokenizers here?

gchhablani replied:
@patrickvonplaten There are two different types of models. One is trained with 7 languages, and one with 3 languages. `MultiTokenizer` is with 7 languages, so we need different language codes for that, hence I have added two tokenizers.

gchhablani replied:
@patil-suraj There is actually another difference. The `MultiTokenizer` is only used for fine-tuning purposes. Hence, it does not have `<mask>` token. Also, in the case of the regular tokenizer, there is a need for allowing no language token, but for `MultiTokenizer` there will always be a language code.
@wasiahmad Please correct me if I am wrong here.

wasiahmad replied:
@gchhablani You are right. As shown [here](https://github.com/wasiahmad/PLBART/blob/main/multilingual/plbart/convert.py), for multilingual fine-tuning, we removed the `<mask>` token and added 4 new language tokens. 
I remember, in HF tokenizer, we can add special tokens using the `additional_special_tokens`, so for the multilingual fine-tuned checkpoint, we may use this parameter to add the language tokens?
Also, is it possible to exclude the `<mask>` token from the tokenizer by setting `mask_token = None`? In such a case, we may combine the two tokenizers. Please correct me if I am wrong, I do not know what happens with the `<mask>` token during downstream task fine-tuning.

gchhablani replied:
@patil-suraj In this case, is it possible to accomodate both of them together? Wdyt?

patil-suraj replied:
Thanks @wasiahmad!
And yes, we can still accommodate this within a single tokenizer. As Wasi said, we can use the `additional_special_tokens` for the language tokens, for ex: that's what `M2M100Tokenizer` does. 

For this, we could add an argument like `use_language_codes` and if it's `True` add language codes else don't add.

Same for the `<mask>` token. Setting `mask_token = None` should work if the mask token is not required.

patrickvonplaten commented on src/transformers/convert_slow_tokenizer.py:
```diff
@@ -881,6 +881,67 @@ def converted(self) -> Tokenizer:
         return tokenizer
 
 
+class PLBartConverter(SpmConverter):
```
Let's take care of this in the very end - I can help you here :-)

patrickvonplaten commented on src/transformers/models/plbart/__init__.py:
```diff
@@ -0,0 +1,70 @@
+# There's no way to ignore "F401 '...' imported but unused" warnings in this
+# module, but to preserve other warnings. So, don't check this module at all.
+
+# Copyright 2020 The HuggingFace Team. All rights reserved.
```
```suggestion
# Copyright 2021 The HuggingFace Team. All rights reserved.
```

patrickvonplaten commented on src/transformers/models/plbart/configuration_plbart.py:
```diff
@@ -0,0 +1,193 @@
+
+    def __init__(
+        self,
+        vocab_size=50005,  # TODO: Check if vocab size includes the language codes?
```
are they pretrained weights? The most important here is to match the pretrained weights

patrickvonplaten replied:
Think we should convert the pretrained weights and then see how large the word embeddings are in the checkpoint

gchhablani replied:
![image](https://user-images.githubusercontent.com/29076344/148259125-d49d1402-2594-4538-b9f0-e57c1e843d98.png)
It is 50005.

patrickvonplaten commented on src/transformers/models/plbart/modeling_plbart.py:
```diff
@@ -0,0 +1,1909 @@
+_TOKENIZER_FOR_DOC = "PLBartTokenizer"
+
+
+PLBART_PRETRAINED_MODEL_ARCHIVE_LIST = [  # TODO: Check if this is the place for fine-tuned ckpts as well.
```
Feel free to add fine-tuned ckpts as well, but there is absolutely no need to do so. A single checkpoint of this model architecture is definitely enough here

gchhablani replied:
I am letting the fine-tuned ckpts be there.

patrickvonplaten commented on src/transformers/models/plbart/modeling_plbart.py:
```diff
@@ -0,0 +1,1909 @@
+        return attn_output, attn_weights_reshaped, past_key_value
+
+
+class PLBartEncoderLayer(nn.Module):
```
This is different from Bart?

patrickvonplaten replied:
or MBart?

gchhablani replied:
It is the same as Bart. Does this also need a `Copied from`?

patrickvonplaten commented on src/transformers/models/plbart/modeling_plbart.py:
```diff
@@ -0,0 +1,1909 @@
+    """,
+    PLBART_START_DOCSTRING,
+)
+class PLBartForSequenceClassification(PLBartPreTrainedModel):
```
Is this needed in a first step? Is there any fine-tuned checkpoint for SequenceClassification?

gchhablani replied:
Yes, there are two checkpoints for `SequenceClassification`.

patrickvonplaten commented on src/transformers/models/plbart/modeling_plbart.py:
```diff
@@ -0,0 +1,1909 @@
+    """,
+    PLBART_START_DOCSTRING,
+)
+class PLBartForQuestionAnswering(PLBartPreTrainedModel):
```
some here - is there already a pretrained checkpoint for this task (QA)?

gchhablani replied:
No, no fine-tuned checkpoints for `QuestionAnswering` model.

patrickvonplaten commented on src/transformers/models/plbart/tokenization_plbart.py:
```diff
@@ -0,0 +1,256 @@
+]
+
+
+class PLBartTokenizer(XLMRobertaTokenizer):
```
Please don't use an abstraction here. We are trying to remove all abstractions from the tokenizers. So it should be just:

```
class PLBartTokenizer(PreTrainedTokenizer):
```

gchhablani replied:
@patrickvonplaten @patil-suraj
What should I do about the inherited methods? Should I just copy all of them here?

patrickvonplaten commented on src/transformers/models/plbart/tokenization_plbart_fast.py:
```diff
@@ -0,0 +1,271 @@
+# coding=utf-8
```
Let's try to tackle this one in a final step once all the comments above are corrected

patrickvonplaten commented on src/transformers/models/plbart_multi/tokenization_plbart_multi.py:
```diff
@@ -0,0 +1,304 @@
+]
+
+
+class PLBartMultiTokenizer(XLMRobertaTokenizer):
```
Is this one really that different that we can't put it into the same class?

gchhablani replied:
The two tokenizers have different language codes.

patrickvonplaten commented on tests/test_modeling_plbart.py:
```diff
@@ -0,0 +1,658 @@
+        config_and_inputs = self.model_tester.prepare_config_and_inputs_for_common()
+        self.model_tester.check_encoder_decoder_model_standalone(*config_and_inputs)
+
+    # PLBartForSequenceClassification does not support inputs_embeds
```
happy to remove `PLBartForSequenceClassification` and `PLBartForQuestionAnswering` in general if there are no pretrained checkpoitns

patrickvonplaten commented on tests/test_tokenization_plbart.py:
```diff
@@ -0,0 +1,278 @@
+class PLBartTokenizationTest(TokenizerTesterMixin, unittest.TestCase):
+    tokenizer_class = PLBartTokenizer
+    rust_tokenizer_class = PLBartTokenizerFast
+    test_rust_tokenizer = True
```
Let's set this flag to `False` in a first step to make sure all the slow tokenizer tests work correctly.

patrickvonplaten reviewed (commented):
Hey @gchhablani,

Thanks for the great model addition! I think the PR is already in a very good state. I'll do my best this week to help you get this PR merged :-) 

Could you in a first step do the following:

1) Merge master into your PR or rebase PR to current master (sorry that it took us so long to review)
2) remove the `XLMRobertaTokenizer` dependency of the PLBart tokenizer.\
3) I think we should try to just have a single tokenizer file instead of two. Could you check whether it's possible to move all tokenizer-relevant code into a single class?
4) Disable the fast tokenizer test (but keep all the code here) 
5) Apply the other suggestions
6) Then we can iterate on the PR until all tests pass
7)  In a final step, I can go into the PR and help you get the fast tokenizer working 

Does that sound good to you? 

Super sorry to have this PR on hold for so long!

patrickvonplaten commented:
@gchhablani - can I help you somehow to move forward with this PR?

gchhablani committed:
Add changes from review

gchhablani committed:
Fix copies

gchhablani committed:
Fix docs

wasiahmad commented:
Any update on this PR? @gchhablani

patil-suraj commented on src/transformers/models/plbart/modeling_plbart.py:
```diff
@@ -0,0 +1,1795 @@
+"""
+
+
+class PLBartEncoder(PLBartPreTrainedModel):
```
Add `Copied from` if possible.

patil-suraj commented on src/transformers/models/plbart/modeling_plbart.py:
```diff
@@ -0,0 +1,1795 @@
+        )
+
+
+class PLBartDecoder(PLBartPreTrainedModel):
```
Add `Copied from` if possible.

patil-suraj commented on src/transformers/models/plbart/tokenization_plbart.py:
```diff
@@ -0,0 +1,254 @@
+# coding=utf-8
+# Copyright 2021, UCLA NLP, The Facebook AI Research Team Authors and The HuggingFace Inc. team.
```
```suggestion
# Copyright 2022, UCLA NLP, The Facebook AI Research Team Authors and The HuggingFace Inc. team.
```

patil-suraj commented on src/transformers/models/plbart/tokenization_plbart.py:
```diff
@@ -0,0 +1,254 @@
+PRETRAINED_VOCAB_FILES_MAP = {
+    "vocab_file": {
+        "uclanlp/plbart-base": "https://huggingface.co/uclanlp/plbart-base/resolve/main/sentencepiece.bpe.model",
+        "uclanlp/plbart-c-cpp-defect-detection": "https://huggingface.co/uclanlp/plbart-c-cpp-defect-detection/resolve/main/sentencepiece.bpe.model",
+        "uclanlp/plbart-cs-java": "https://huggingface.co/uclanlp/plbart-cs-java/resolve/main/sentencepiece.bpe.model",
+        "uclanlp/plbart-en_XX-java": "https://huggingface.co/uclanlp/plbart-en_XX-java/resolve/main/sentencepiece.bpe.model",
+        "uclanlp/plbart-go-en_XX": "https://huggingface.co/uclanlp/plbart-go-en_XX/resolve/main/sentencepiece.bpe.model",
+        "uclanlp/plbart-java-clone-detection": "https://huggingface.co/uclanlp/plbart-java-clone-detection/resolve/main/sentencepiece.bpe.model",
+        "uclanlp/plbart-java-cs": "https://huggingface.co/uclanlp/plbart-java-cs/resolve/main/sentencepiece.bpe.model",
+        "uclanlp/plbart-java-en_XX": "https://huggingface.co/uclanlp/plbart-java-en_XX/resolve/main/sentencepiece.bpe.model",
+        "uclanlp/plbart-javascript-en_XX": "https://huggingface.co/uclanlp/plbart-javascript-en_XX/resolve/main/sentencepiece.bpe.model",
+        "uclanlp/plbart-php-en_XX": "https://huggingface.co/uclanlp/plbart-php-en_XX/resolve/main/sentencepiece.bpe.model",
+        "uclanlp/plbart-python-en_XX": "https://huggingface.co/uclanlp/plbart-python-en_XX/resolve/main/sentencepiece.bpe.model",
+        "uclanlp/plbart-refine-java-medium": "https://huggingface.co/uclanlp/plbart-refine-java-medium/resolve/main/sentencepiece.bpe.model",
+        "uclanlp/plbart-refine-java-small": "https://huggingface.co/uclanlp/plbart-refine-java-small/resolve/main/sentencepiece.bpe.model",
+        "uclanlp/plbart-ruby-en_XX": "https://huggingface.co/uclanlp/plbart-ruby-en_XX/resolve/main/sentencepiece.bpe.model",
```
Not necessary to have all checkpoints here.

gchhablani replied:
I think for different tokenizers, it will be better to specify which to use for which model to avoid confusion. Wdyt?

I'll also prepare a table and maybe add to the docs.

patil-suraj commented on src/transformers/models/plbart/tokenization_plbart.py:
```diff
@@ -0,0 +1,254 @@
+}
+
+PRETRAINED_POSITIONAL_EMBEDDINGS_SIZES = {
+    "uclanlp/plbart-base": 1024,
+    "uclanlp/plbart-c-cpp-defect-detection": 1024,
+    "uclanlp/plbart-cs-java": 1024,
+    "uclanlp/plbart-en_XX-java": 1024,
+    "uclanlp/plbart-go-en_XX": 1024,
+    "uclanlp/plbart-java-clone-detection": 1024,
+    "uclanlp/plbart-java-cs": 1024,
+    "uclanlp/plbart-java-en_XX": 1024,
+    "uclanlp/plbart-javascript-en_XX": 1024,
+    "uclanlp/plbart-php-en_XX": 1024,
+    "uclanlp/plbart-python-en_XX": 1024,
+    "uclanlp/plbart-refine-java-medium": 1024,
+    "uclanlp/plbart-refine-java-small": 1024,
+    "uclanlp/plbart-ruby-en_XX": 1024,
```
same comment as above

patil-suraj commented on src/transformers/models/plbart_multi/__init__.py:
```diff
@@ -0,0 +1,42 @@
+# flake8: noqa
+# There's no way to ignore "F401 '...' imported but unused" warnings in this
+# module, but to preserve other warnings. So, don't check this module at all.
+
+# Copyright 2020 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
```
This file should be removed IMO since `plbart_multi` is not a different model.

patil-suraj commented on tests/test_modeling_plbart.py:
```diff
@@ -0,0 +1,655 @@
+# coding=utf-8
+# Copyright 2021, The HuggingFace Inc. team. All rights reserved.
```
```suggestion
# Copyright 2022, The HuggingFace Inc. team. All rights reserved.
```

patil-suraj commented on tests/test_modeling_plbart.py:
```diff
@@ -0,0 +1,655 @@
+            encoder.save_pretrained(tmpdirname)
+            encoder = PLBartEncoder.from_pretrained(tmpdirname).to(torch_device)
+
+        encoder_last_hidden_state_2 = encoder(inputs_dict["input_ids"], attention_mask=inputs_dict["attention_mask"])[
+            0
+        ]
```
hard to read, split on two lines

patil-suraj commented on tests/test_modeling_plbart.py:
```diff
@@ -0,0 +1,655 @@
+    return torch.tensor(tok_lst, dtype=torch.long, device=torch_device)
+
+
+@require_torch
+@require_sentencepiece
+@require_tokenizers
+class AbstractSeq2SeqIntegrationTest(unittest.TestCase):
+    maxDiff = 1000  # longer string compare tracebacks
+    checkpoint_name = None
+
+    @classmethod
+    def setUpClass(cls):
+        cls.tokenizer = AutoTokenizer.from_pretrained(cls.checkpoint_name, use_fast=False)
+        return cls
+
+    @cached_property
+    def model(self):
+        """Only load the model if needed."""
+        model = PLBartForConditionalGeneration.from_pretrained(self.checkpoint_name).to(torch_device)
+        if "cuda" in torch_device:
+            model = model.half()
+        return model
+
```
Nice tests!

patil-suraj commented on tests/test_tokenization_plbart.py:
```diff
@@ -0,0 +1,278 @@
+# Copyright 2020 The HuggingFace Team. All rights reserved.
```
```suggestion
# Copyright 2022 The HuggingFace Team. All rights reserved.
```

patil-suraj reviewed (commented):
Great work @gchhablani , it's looking good! Sorry about the super late review.

I have pretty much the same comments as Patrick's about the two tokenizers and tokenizer inheritance. 
Apart from that:
- Let's try to have just one tokenizer, I've left a comment about how this can be done.
- Let's rebase to master and make sure all docstrings are in markdown instead of `rst`. Also left a comment about this.
- Use `Copied from`  wherever applicable.

We should be good to merge once this is done.
Let me know if you need help :)

gchhablani commented on docs/source/model_doc/plbart.rst:
```diff
@@ -0,0 +1,83 @@
+    an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+    specific language governing permissions and limitations under the License.
+
+PLBART
+-----------------------------------------------------------------------------------------------------------------------
+
+Overview
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The PLBART model was proposed in `Unified Pre-training for Program Understanding and Generation <https://arxiv.org/abs/2103.06333>`__ by Wasi Uddin Ahmad, Saikat Chakraborty, Baishakhi Ray, Kai-Wei Chang.
+This is a BART-like model which can be used to perform code-summarization, code-generation, and code-translation tasks. The pre-trained model `plbart-base` has been trained using multilingual denoising task
+on Java, Python and English.
+
+
+The abstract from the paper is the following:
+
```
@patil-suraj I have tried using `doc-builder` but for some reason it didn't convert the `rst` docs to `mdx`. So I created my own. Hope that is cool?

gchhablani committed:
Fix docs

gchhablani committed:
Fix style

gchhablani commented on src/transformers/models/plbart/configuration_plbart.py:
```diff
@@ -0,0 +1,191 @@
+
+class PLBartConfig(PretrainedConfig):
+    r"""
+    This is the configuration class to store the configuration of a :class:`~transformers.PLBartModel`. It is used to
+    instantiate an PLBART model according to the specified arguments, defining the model architecture. Instantiating a
+    configuration with the defaults will yield a similar configuration to that of the PLBART `uclanlp/plbart-base
+    <https://huggingface.co/uclanlp/plbart-base>`__ architecture.
+
+    Configuration objects inherit from :class:`~transformers.PretrainedConfig` and can be used to control the model
+    outputs. Read the documentation from :class:`~transformers.PretrainedConfig` for more information.
+
+
+    Args:
+        vocab_size (:obj:`int`, `optional`, defaults to 50005):
+            Vocabulary size of the PLBART model. Defines the number of different tokens that can be represented by the
+            :obj:`inputs_ids` passed when calling :class:`~transformers.PLBartModel`.
+        d_model (:obj:`int`, `optional`, defaults to 768):
+            Dimensionality of the layers and the pooler layer.
+        encoder_layers (:obj:`int`, `optional`, defaults to 6):
+            Number of encoder layers.
+        decoder_layers (:obj:`int`, `optional`, defaults to 6):
+            Number of decoder layers.
+        encoder_attention_heads (:obj:`int`, `optional`, defaults to 12):
+            Number of attention heads for each attention layer in the Transformer encoder.
+        decoder_attention_heads (:obj:`int`, `optional`, defaults to 12):
+            Number of attention heads for each attention layer in the Transformer decoder.
+        decoder_ffn_dim (:obj:`int`, `optional`, defaults to 3072):
+            Dimensionality of the "intermediate" (often named feed-forward) layer in decoder.
+        encoder_ffn_dim (:obj:`int`, `optional`, defaults to 3072):
```
I have created the `mdx` file. Will it fix the problem?

gchhablani committed:
Fix year

gchhablani commented on src/transformers/models/plbart/modeling_plbart.py:
```diff
@@ -0,0 +1,1795 @@
+        model(input_ids).logits
+
+        >>> masked_index = (input_ids[0] == tokenizer.mask_token_id).nonzero().item() >>> probs = logits[0,
+        masked_index].softmax(dim=0) >>> values, predictions = probs.topk(5)
```
Should everything from here be added to the `mdx`? Should I remove documentation from all the files?

gchhablani replied:
Wouldn't the docstrings be picked up in the `mdx` with `[[autodoc]]`?

gchhablani commented on src/transformers/models/plbart/modeling_plbart.py:
```diff
@@ -0,0 +1,1795 @@
+        )
+        self.layers = nn.ModuleList([PLBartEncoderLayer(config) for _ in range(config.encoder_layers)])
+        self.layernorm_embedding = nn.LayerNorm(embed_dim)
+        # self.layer_norm = nn.LayerNorm(config.d_model) # TODO: Check if this is okay
```
The commented one? No. I think it is a vestige from a previous commit.

gchhablani committed:
Add changes from review

gchhablani committed:
Remove extra changes

gchhablani commented:
@patil-suraj
Updates:
- Rebased the branch to master.
- Updated the documentation. Should I move the generation example and other model/tokenizer docs from the model/tokenization files to the `mdx`?
- Removed all the extra changes due to merging issues.

Regarding the tokenizes, I am not sure if they can be moved to a single tokenizer considering [this comment](https://github.com/huggingface/transformers/pull/13269#discussion_r790167690). Please let me know if there's a way to do that.

gchhablani committed:
Fix base tokenizer and doc

gchhablani committed:
Fix style

gchhablani committed:
Fix modeling and slow tokenizer tests

gchhablani committed:
Remove Multi-tokenizer Converter and Tests

gchhablani committed:
Delete QA model and Multi Tokenizer dummy objects

gchhablani committed:
Fix repo consistency and code quality issues

gchhablani committed:
Fix example documentation

gchhablani committed:
Fix style

gchhablani committed:
Remove PLBartTokenizer from type checking in init

gchhablani committed:
Fix consistency issue

patil-suraj commented on src/transformers/models/plbart/configuration_plbart.py:
```diff
@@ -0,0 +1,192 @@
+        decoder_layers=6,
+        decoder_ffn_dim=3072,
+        decoder_attention_heads=12,
+        encoder_layerdrop=0.0,  # TODO: Check
+        decoder_layerdrop=0.0,  # TODO: Check
```
should check and remove this comment.

patil-suraj commented on src/transformers/models/plbart/configuration_plbart.py:
```diff
@@ -0,0 +1,192 @@
+        attention_dropout=0.1,
+        activation_dropout=0.0,
+        init_std=0.02,
+        classifier_dropout=0.0,  # TODO: Verify
```
same comment as above.

patil-suraj commented on src/transformers/models/plbart/tokenization_plbart.py:
```diff
@@ -0,0 +1,450 @@
+            save_directory, (filename_prefix + "-" if filename_prefix else "") + VOCAB_FILES_NAMES["vocab_file"]
+        )
+
+        if os.path.abspath(self.vocab_file) != os.path.abspath(out_vocab_file):
+            copyfile(self.vocab_file, out_vocab_file)
```
```suggestion
        if os.path.abspath(self.vocab_file) != os.path.abspath(out_vocab_file) and os.path.isfile(self.vocab_file):
            copyfile(self.vocab_file, out_vocab_file)
        elif not os.path.isfile(self.vocab_file):
            with open(out_vocab_file, "wb") as fi:
                content_spiece_model = self.sp_model.serialized_model_proto()
                fi.write(content_spiece_model)
```

patil-suraj commented on src/transformers/models/plbart/tokenization_plbart_fast.py:
```diff
@@ -0,0 +1,275 @@
+]
+
+
+class PLBartTokenizerFast(XLMRobertaTokenizerFast):
```
Let's not use inheritance here, should add this as a standalone tokenizer.

gchhablani replied:
@patil-suraj 
Removing fast tokenizer for now.

patil-suraj commented on src/transformers/models/plbart/tokenization_plbart_multi.py:
```diff
@@ -0,0 +1,307 @@
+# coding=utf-8
+# Copyright 2020 The Facebook AI Research Team Authors and The HuggingFace Inc. team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from contextlib import contextmanager
+from typing import List, Optional
+
+from ...tokenization_utils import BatchEncoding
+from ...utils import logging
+from ..xlm_roberta.tokenization_xlm_roberta import XLMRobertaTokenizer
+
+
+logger = logging.get_logger(__name__)
+
+
+VOCAB_FILES_NAMES = {"vocab_file": "sentencepiece.bpe.model", "tokenizer_file": "tokenizer.json"}
```
We should delete this file

patil-suraj commented on src/transformers/models/plbart/tokenization_plbart_multi_fast.py:
```diff
@@ -0,0 +1,348 @@
+# coding=utf-8
+# Copyright 2020 The Facebook AI Research Team Authors and The HuggingFace Inc. team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from contextlib import contextmanager
+from typing import List, Optional
+
+from tokenizers import processors
+
+from ...file_utils import is_sentencepiece_available
+from ...tokenization_utils import BatchEncoding
+from ...utils import logging
```
We should delete this file.

patil-suraj commented on docs/source/model_doc/plbart.mdx:
```diff
@@ -0,0 +1,121 @@
+
+[[autodoc]] PLBartTokenizerFast
+
+## PLBart50Tokenizer
+
+[[autodoc]] PLBart50Tokenizer
+
+## PLBart50TokenizerFast
+
+[[autodoc]] PLBart50TokenizerFast
```
No PLBart50Tokenizer tokenizer
```suggestion
```

patil-suraj reviewed (approved):
The looks ready now, thanks a lot for finishing this!
I left a few comments, which should help resolve the failing tests. After that we should be good to merge :)

gchhablani committed:
Add changes from review

gchhablani committed:
Fix style

patil-suraj commented on docs/source/model_doc/plbart.mdx:
```diff
@@ -0,0 +1,113 @@
+    - as_target_tokenizer
+    - build_inputs_with_special_tokens
+
+## PLBartTokenizerFast
+
+[[autodoc]] PLBartTokenizerFast
```
this should be deleted since we removed the fast tokenizer.

patil-suraj commented on src/transformers/__init__.py:
```diff
@@ -425,6 +427,7 @@
 # tokenizers-backed objects
 if is_tokenizers_available():
     # Fast tokenizers
+    _import_structure["models.plbart"].append("PLBartTokenizerFast")
```
same comment as above

patil-suraj commented on src/transformers/__init__.py:
```diff
@@ -2618,6 +2633,7 @@
         from .models.mt5 import MT5TokenizerFast
         from .models.openai import OpenAIGPTTokenizerFast
         from .models.pegasus import PegasusTokenizerFast
+        from .models.plbart import PLBartTokenizerFast
```
same comment as above

patil-suraj commented on src/transformers/models/auto/tokenization_auto.py:
```diff
@@ -47,6 +47,13 @@
+                "plbart",
+                (
+                    "PLBartTokenizer" if is_sentencepiece_available() else None,
+                    "PLBartTokenizerFast" if is_tokenizers_available() else None,
```
same comment as above

patil-suraj commented on src/transformers/models/plbart/__init__.py:
```diff
@@ -0,0 +1,68 @@
+    _import_structure["tokenization_plbart"] = ["PLBartTokenizer"]
+
+
+if is_tokenizers_available():
+    _import_structure["tokenization_plbart_fast"] = ["PLBartTokenizerFast"]
```
same comment as above

patil-suraj commented on src/transformers/models/plbart/__init__.py:
```diff
@@ -0,0 +1,68 @@
+    if is_sentencepiece_available():
+        from .tokenization_plbart import PLBartTokenizer
+
+    if is_tokenizers_available():
+        from .tokenization_plbart_fast import PLBartTokenizerFast
```
same comment as above

gchhablani committed:
Remove PLBartTokenizerFast

patil-suraj commented on src/transformers/convert_slow_tokenizer.py:
```diff
@@ -910,6 +910,34 @@ def converted(self) -> Tokenizer:
         return tokenizer
 
 
+class PLBartConverter(SpmConverter):
```
This should also be removed.

patil-suraj commented on src/transformers/convert_slow_tokenizer.py:
```diff
@@ -971,6 +999,7 @@ def post_processor(self):
     "MobileBertTokenizer": BertConverter,
     "OpenAIGPTTokenizer": OpenAIGPTConverter,
     "PegasusTokenizer": PegasusConverter,
+    "PLBartTokenizer": PLBartConverter,
```
same comment as above.

gchhablani committed:
Remove FastTokenizer converter

gchhablani committed:
Fix AutoTokenzier mapping

gchhablani committed:
Add plbart to toctree and fix consistency issues

gchhablani committed:
Add language codes tokenizer test

patil-suraj commented:
@patrickvonplaten Would be awesome if you could take a final look now :)

patrickvonplaten commented on tests/test_modeling_plbart.py:
```diff
@@ -0,0 +1,655 @@
+
+
+@require_torch
+class PLBartStandaloneDecoderModelTest(ModelTesterMixin, GenerationTesterMixin, unittest.TestCase):
```
Very nice to have left this here as well

gchhablani replied:
@patrickvonplaten Should I create a copy of this test for `PLBartForCasualLM`?

patrickvonplaten replied:
Nono it's perfect :-)

patrickvonplaten commented:
PR is good for merge to me!

sgugger commented on src/transformers/models/plbart/modeling_plbart.py:
```diff
@@ -0,0 +1,1734 @@
+        Args:
+            hidden_states (`torch.FloatTensor`): input to the layer of shape *(seq_len, batch, embed_dim)*
+            attention_mask (`torch.FloatTensor`): attention mask of size
+                *(batch, 1, tgt_len, src_len)* where padding elements are indicated by very large negative values.
```
The original docstring in `BartEncoderLayer` needs fixing here and below, the shapes should be between \`.

sgugger commented on src/transformers/models/plbart/modeling_plbart.py:
```diff
@@ -0,0 +1,1734 @@
+        Args:
+            hidden_states (`torch.FloatTensor`): input to the layer of shape *(batch, seq_len, embed_dim)*
+            attention_mask (`torch.FloatTensor`): attention mask of size
+                *(batch, 1, tgt_len, src_len)* where padding elements are indicated by very large negative values.
```
Same here.

sgugger commented on src/transformers/models/plbart/modeling_plbart.py:
```diff
@@ -0,0 +1,1795 @@
+        if isinstance(module, (PLBartDecoder, PLBartDecoder)):
+            module.gradient_checkpointing = value
+
+    @property
+    def dummy_inputs(self):
+        pad_token = self.config.pad_token_id
+        input_ids = torch.tensor([[0, 6, 10, 4, 2], [0, 8, 12, 2, pad_token]], device=self.device)
+        dummy_inputs = {
+            "attention_mask": input_ids.ne(pad_token),
+            "input_ids": input_ids,
+        }
+        return dummy_inputs
```
This has been resolved but not addressed.

sgugger commented on src/transformers/models/plbart/modeling_plbart.py:
```diff
@@ -0,0 +1,1734 @@
+            "attention_mask": input_ids.ne(pad_token),
+            "input_ids": input_ids,
+        }
+        return dummy_inputs
```
GitHub is not letting me unresolve the comment, so adding a new one here. The dummy_inputs method should be removed.

sgugger commented on src/transformers/models/plbart/modeling_plbart.py:
```diff
@@ -0,0 +1,1734 @@
+"""
+
+PLBART_GENERATION_EXAMPLE = r"""
+    Token in-filling example::
+
+        >>> from transformers import PLBartTokenizer, PLBartForConditionalGeneration, PLBartConfig
+
+        >>> model = PLBartForConditionalGeneration.from_pretrained('uclanlp/plbart-base') >>> tokenizer =
+        PLBartTokenizer.from_pretrained('uclanlp/plbart-base', src_lang='java', tgt_lang='java')
+
+        >>> METHOD_TO_FILL = "public static main (String args[0]) { data=Date(); System.out. String.format("Current
+        Date : % tc", ));}" >>> inputs = tokenizer([METHOD_TO_FILL], max_length=1024, return_tensors='pt')
+
+        >>> # Generate Filled Code >>> generated_ids = model.generate(inputs['input_ids'], num_beams=4, max_length=5,
+        early_stopping=True) >>> print([tokenizer.decode(g, skip_special_tokens=True,
+        clean_up_tokenization_spaces=False) for g in generated_ids])
+
+    Mask-filling example::
+
+        >>> from transformers import PLBartTokenizer, PLBartForConditionalGeneration >>> tokenizer =
+        PLBartTokenizer.from_pretrained('uclanlp/plbart-base') >>> # en_XX is the language symbol id <LID> for English
+        >>> TXT = "</s> Is 0 the <mask> Fibonacci <mask> ? </s> en_XX"
+
+        >>> model = PLBartForConditionalGeneration.from_pretrained('uclanlp/plbart-base') >>> input_ids =
+        tokenizer([TXT], add_special_tokens=False, return_tensors='pt')['input_ids'] >>> logits =
+        model(input_ids).logits
+
+        >>> masked_index = (input_ids[0] == tokenizer.mask_token_id).nonzero().item() >>> probs = logits[0,
+        masked_index].softmax(dim=0) >>> values, predictions = probs.topk(5)
```
Those examples need fixing, and tehre should only be one colon after example. We don't use rst anymore.

sgugger commented on src/transformers/models/plbart/modeling_plbart.py:
```diff
@@ -0,0 +1,1734 @@
+            self.embed_tokens = nn.Embedding(config.vocab_size, config.d_model, self.padding_idx)
+
+        self.embed_positions = PLBartLearnedPositionalEmbedding(
+            config.max_position_embeddings,
+            config.d_model,
```
```suggestion
            config.max_position_embeddings, config.d_model
```
This cna even probably fit in one line with the embedding class above.

sgugger commented on src/transformers/models/plbart/tokenization_plbart.py:
```diff
@@ -0,0 +1,454 @@
+
+FAIRSEQ_LANGUAGE_CODES = {
+    "base": ["java", "python", "en_XX"],
+    "multi": [
+        "java",
+        "python",
+        "en_XX",
+        "javascript",
+        "php",
+        "ruby",
+        "go",
+    ],
```
This should git in one line.

sgugger commented on src/transformers/models/plbart/tokenization_plbart.py:
```diff
@@ -0,0 +1,454 @@
+    """
+    Construct an PLBART tokenizer.
+
+    Adapted from [*RobertaTokenizer*] and [*XLNetTokenizer*]. Based on
+    [SentencePiece](https://github.com/google/sentencepiece).
```
```suggestion
    Adapted from [`RobertaTokenizer`] and [`XLNetTokenizer`]. Based on
    [SentencePiece](https://github.com/google/sentencepiece).
```

sgugger commented on src/transformers/models/plbart/tokenization_plbart.py:
````diff
@@ -0,0 +1,454 @@
+    Adapted from [*RobertaTokenizer*] and [*XLNetTokenizer*]. Based on
+    [SentencePiece](https://github.com/google/sentencepiece).
+
+    The tokenization method is *<tokens> <eos> <language code>* for source language documents, and ``<language code>
+    <tokens> <eos>``` for target language documents.
````
```suggestion
    The tokenization method is `<tokens> <eos> <language code>` for source language documents, and `<language code>
    <tokens> <eos>` for target language documents.
```

sgugger commented on src/transformers/models/plbart/tokenization_plbart.py:
```diff
@@ -0,0 +1,454 @@
+        self, token_ids_0: List[int], token_ids_1: Optional[List[int]] = None
+    ) -> List[int]:
+        """
+        Args:
+        Create a mask from the two sequences passed to be used in a sequence-pair classification task. PLBart does not
+        make use of token type ids, therefore a list of zeros is returned. TODO: Check if this is correct.
```
```suggestion
        Create a mask from the two sequences passed to be used in a sequence-pair classification task. PLBart does not
        make use of token type ids, therefore a list of zeros is returned. TODO: Check if this is correct.

        Args:
```
Also the TODO should be addressed and removed.

gchhablani replied:
The model is same as Bart, so I believe tokenization will work in the same fashion. Removing the `TO-DO`.

sgugger commented on tests/test_modeling_plbart.py:
```diff
@@ -0,0 +1,655 @@
+
+    def prepare_config_and_inputs(self):
+        input_ids = ids_tensor([self.batch_size, self.seq_length], self.vocab_size)
+        input_ids = ids_tensor([self.batch_size, self.seq_length], self.vocab_size).clamp(
+            3,
+        )
```
The clamp should be done on a new line to be more readable.

sgugger commented on tests/test_modeling_plbart.py:
```diff
@@ -0,0 +1,655 @@
+
+        output_from_no_past = model(next_input_ids, attention_mask=next_attention_mask)["last_hidden_state"]
+        output_from_past = model(next_tokens, attention_mask=next_attention_mask, past_key_values=past_key_values)[
+            "last_hidden_state"
+        ]
```
Selecting the last hidden state should be done on a separate line to be mroe readable.

sgugger commented on tests/test_modeling_plbart.py:
```diff
@@ -0,0 +1,655 @@
+
+    @slow
+    def test_java_cs_generate_one(self):
+        batch: BatchEncoding = self.tokenizer(
+            ["public int maximum(int a, int b, int c){return Math.max(a, Math.max(b, c));}"], return_tensors="pt"
+        ).to(torch_device)
```
No need for the type annotation here.
```suggestion
        batch = self.tokenizer(
            ["public int maximum(int a, int b, int c){return Math.max(a, Math.max(b, c));}"], return_tensors="pt"
        ).to(torch_device)
```

sgugger commented on tests/test_modeling_plbart.py:
```diff
@@ -0,0 +1,655 @@
+
+    @slow
+    def test_java_cs_generate_batch(self):
+        batch: BatchEncoding = self.tokenizer(self.src_text, return_tensors="pt", padding=True, truncation=True).to(
```
Same here.
Pluse the `to` should be done on its own line to be more readable.

sgugger commented on tests/test_modeling_plbart.py:
```diff
@@ -0,0 +1,655 @@
+
+    def test_plbart_java_cs_config(self):
+        plbart_models = ["uclanlp/plbart-java-cs"]
+        expected = {
+            "scale_embedding": True
+        }  # TODO: Check if "output_past": True is needed. Currently throws an error.
```
What's the status of the TODO? It should be put on a separate line to be able to have the expected dict in one line.

gchhablani replied:
Currently, the `plbart-java-cs` model does not have `output_past` attribute in its configuration. Removing the `TODO`.

sgugger commented on tests/test_modeling_plbart.py:
```diff
@@ -0,0 +1,655 @@
+@require_tokenizers
+class PLBartBaseIntegrationTest(AbstractSeq2SeqIntegrationTest):
+    checkpoint_name = "uclanlp/plbart-base"
+    src_text = [
+        "Is 0 the first Fibonacci number ?",
+        "Find the sum of all prime numbers .",
+    ]
```
This fits in one line.

sgugger commented on tests/test_modeling_plbart.py:
```diff
@@ -0,0 +1,655 @@
+            is_encoder_decoder=self.is_encoder_decoder,
+        )
+
+        return (
+            config,
+            input_ids,
+            attention_mask,
+            lm_labels,
+        )
```
This fits in one line.

sgugger commented on tests/test_modeling_plbart.py:
```diff
@@ -0,0 +1,655 @@
+        output_from_past_slice = output_from_past[:, 0, random_slice_idx].detach()
+
+        # test that outputs are equal for slice
+        assert torch.allclose(output_from_past_slice, output_from_no_past_slice, atol=1e-3)
```
Use a `self.assertTrue` here.

sgugger commented on tests/test_modeling_plbart.py:
```diff
@@ -0,0 +1,655 @@
+
+    def prepare_config_and_inputs_for_common(self):
+        config_and_inputs = self.prepare_config_and_inputs()
+        (
+            config,
+            input_ids,
+            attention_mask,
+            lm_labels,
+        ) = config_and_inputs
```
This fits in one line.

sgugger commented on tests/test_modeling_plbart.py:
```diff
@@ -0,0 +1,655 @@
+            lm_labels,
+        ) = config_and_inputs
+
+        inputs_dict = {
+            "input_ids": input_ids,
+            "attention_mask": attention_mask,
+        }
```
This too.

sgugger commented on tests/test_modeling_plbart.py:
```diff
@@ -0,0 +1,655 @@
+    test_pruning = False
+    is_encoder_decoder = False
+
+    def setUp(
+        self,
+    ):
```
This fits in one line.

sgugger commented on tests/test_tokenization_plbart.py:
```diff
@@ -0,0 +1,377 @@
+            [
+                value + tokenizer.fairseq_offset
+                for value in [8, 21, 84, 55, 24, 19, 7, 2, 602, 347, 347, 347, 3, 12, 66, 46, 72, 80, 6, 2, 4]
+                #                                       ^ unk: 2 + 1 = 3                  unk: 2 + 1 = 3 ^
```
This comments makes no sense to me, as it points to a 2.

gchhablani replied:
I'll review this one and get back.

sgugger commented on tests/test_tokenization_plbart.py:
```diff
@@ -0,0 +1,377 @@
+        end = tokenizer.vocab_size
+        language_tokens = [tokenizer.convert_ids_to_tokens(x) for x in range(end - 4, end)]
+
+        self.assertListEqual(
+            language_tokens,
+            ["java", "python", "en_XX", "<mask>"],
+        )
```
Fits in one line.

sgugger commented on tests/test_tokenization_plbart.py:
```diff
@@ -0,0 +1,377 @@
+            [
+                value + tokenizer.fairseq_offset
+                for value in [8, 21, 84, 55, 24, 19, 7, 2, 602, 347, 347, 347, 3, 12, 66, 46, 72, 80, 6, 2, 4]
+                #                                       ^ unk: 2 + 1 = 3                  unk: 2 + 1 = 3 ^
```
Same here.

sgugger commented on tests/test_tokenization_plbart.py:
```diff
@@ -0,0 +1,377 @@
+
+        self.assertListEqual(
+            language_tokens,
+            [
+                "java",
+                "python",
+                "en_XX",
+                "javascript",
+                "php",
+                "ruby",
+                "go",
+            ],
```
This list fits in one line.

sgugger commented on tests/test_tokenization_plbart.py:
```diff
@@ -0,0 +1,377 @@
+
+    def test_python_en_tokenizer_truncation(self):
+        src_text = ["def sum(a,b,c):NEW_LINE_INDENTreturn sum([a,b,c])" * 20]
+        assert isinstance(src_text[0], str)
```
Use `self.assertIsInstance` here.

sgugger commented on tests/test_tokenization_plbart.py:
```diff
@@ -0,0 +1,377 @@
+        batch["decoder_input_ids"] = shift_tokens_right(labels, self.tokenizer.pad_token_id).tolist()
+
+        # fairseq batch: https://gist.github.com/sshleifer/cba08bc2109361a74ac3760a7e30e4f4
+        assert batch.input_ids[1][-2:] == [2, PYTHON_CODE]
+        assert batch.decoder_input_ids[1][0] == EN_CODE
+        assert batch.decoder_input_ids[1][-1] == 2
+        assert labels[1][-2:].tolist() == [2, EN_CODE]
```
Those should be tested with `self.assertEqual`

sgugger reviewed (approved):
Thanks a lot for your work on this PR! There are few things left to clean up, but it's almost ready to be merged!

gchhablani committed:
Fix styling and doc issues

gchhablani committed:
Merge remote-tracking branch 'upstream/master' into add_plbart

gchhablani committed:
Add fixes for failing tests

gchhablani committed:
Merge remote-tracking branch 'upstream/master' into add_plbart

gchhablani committed:
Fix copies

gchhablani committed:
Fix failing modeling test

gchhablani committed:
Change assert to assertTrue in modeling tests

gchhablani committed:
Merge remote-tracking branch 'upstream/master' into add_plbart

patil-suraj commented:
Thanks a lot @gchhablani for all your work on this! Great job!

ydshieh commented:
Hi, @gchhablani @patil-suraj @sgugger 

Looks like `README.md` (and therefore `doc/source/index.mdx`) is not updated for this new added model.
Ideally, it is usually updated when a new model is added, right?

sgugger commented:
Thanks for flagging @ydshieh , we indeed have forgotten to ask for it during our reviews! @gchhablani would you like to add it in a follow-up PR?

gchhablani commented:
@ydshieh Thanks a lot for informing! On it.
